### ISSUE TITLE
Add OpenStack Hypervisor Anti-Affinity

### ISSUE THREAD

**Issue by afritzler** [gardener/machine-controller-manager#30]
Enable soft hypervisor anti-affinity for the OpenStack machine class [1]. That way, machines will be placed on different hypervisors when possible.  

[1] https://github.com/gophercloud/gophercloud/blob/master/acceptance/openstack/compute/v2/compute.go#L447

**Comment by vpnachev** [comment]
@hardikdr @prashanth26 any updates on this?

**Comment by hardikdr** [comment]
No, we haven't really spent time on this issue. Is there anything concrete around what needs to be done and also the requirements?

**Comment by afritzler** [comment]
The background for this feature request is, that theoretically you might run into a situation where 2 or more machines get placed on the same hypervisor. If this hypervisor goes down, you might lose all replicas at the same time.

**Comment by dkistner** [comment]
We would need to create `servergroup(s)` where we attach the machines to. Servergroups have a policy assigned which define the affinity or anti-affinity behaviour of the group. See [terraform docs for servergroups](https://www.terraform.io/docs/providers/openstack/r/compute_servergroup_v2.html#policies-1). 

We should use `soft anti-affinity` as it will try to distribute the machines across different hypervisors wherever possible. In contrast to `hard anti-affinity` which will fail if the machines can't be distributed across different hypervisors (e.g. not sufficient capacity of the hypervisor, more machines then available hypervisors etc.).

I would also suggest to create for each workerpool/machineclass an own servergroup managed by MCM. MCM could simply check if a servergroup for the machineclass already exists and if not it could create it. New machines will be then attached to the servergroup. As existing machines can't be attached to the servergroup after creation a rolling update of the machines belonging to a workerpool/machineclass would be required.

We should definitely test how servergroups behave when machines get frequently attached and detached (k8s node scale out/in). Furthermore we need to ensure that disks can safely move from one hypervisor to another as a servergroup with `soft anti-affinity` will always try to place new machines on different hypervisors and most likely workloads + their disks need to move to new machines on different hypervisors.

cc @vlerenc, @afritzler, @RaphaelVogel

**Comment by vlerenc** [comment]
Thanks for opening @dkistner. FYI: @hardikdr @prashanth26 @amshuman-kr this was feedback we received from Converged Cloud on our VM scheduling practice. From what we understand, they would be happy if it can be implemented by Q320, I think. Is that right @kistner? That would allow us to postpone this thing until we are done with the out-of-tree implementation. Then again, if someone else could pick it up, then it can be done also earlier. 

@dkistner The OpenStack implementation wasn't done by the MCM colleagues themselves. We need to see, whether the MCM can help or someone else would be able to help.

**Comment by kayrus** [comment]
> I would also suggest to create for each workerpool/machineclass an own servergroup managed by MCM

Don't forget to consider server-groups and server-group-members quota. By default it is 10/40 in ccee.

> their disks need to move to new machines on different hypervisors.

We're using iSCSI, so disks can be attached without any restrictions.

The only concern, that sometimes, when VMs are scheduled in parallel, they can end on the same hypervizor independently on the anti-affinity rule. cc @joker-official

**Comment by dkistner** [comment]
@vlerenc I have already implemented similar logic for the multiple availability set story on Azure. So I think it can be adapted with low effort. The main effort is to test. I can do it, but we need to properly prioritise. 

@kayrus Thanks for the hint with the `server-group-member-quota`. I was not aware of that. That would mean we would be forced to have a dedicated server group for worker pool and the size of a worker pool need to be limited?
